File: UOA/PCKWU/1046/main.cpp
#include <bits/stdc++.h>

using namespace std;

template <typename T>
using vec1 = vector<T>;

template <typename T>
using vec2 = vec1<vec1<T>>;

template <typename T>
vec2<T> InitVec2(size_t x, size_t y, const T& i = T())
{
    return vec2<T>(y, vec1<T>(x, i));
}

template <typename T>
T Get(const vec2<T>& v2, size_t x, size_t y)
{
    return v2[y][x];
}

template <typename T>
T& Ref(vec2<T>& v2, size_t x, size_t y)
{
    return v2[y][x];
}

template <typename T>
bool Find(const vec2<T>& v2, const T& val, size_t& x, size_t& y)
{
    for (size_t j = 0; j < v2.size(); ++j) {
        const auto& v1(v2[j]);
        for (size_t i = 0; i < v1.size(); ++i) {
            if (v1[i] == val) {
                x = i;
                y = j;
                return true;
            }
        }
    }
    return false;
}

template <typename T>
T cin2var()
{
    T val;
    cin >> val;
    return val;
}

template <typename T>
vec1<T> cin2vec(size_t size)
{
    vec1<T> vec1;
    for (size_t i = 0; i < size; ++i) {
        vec1.push_back(cin2var<T>());
    }
    return vec1;
}

constexpr int Inf = INT_MAX / 2;

struct Immutable
{
    size_t     W;
    size_t     H;
    vec2<char> vec2map;
    string     pattern;
};

struct Mutable
{
    vec2<int> vec2map;
    int       max_count;
};

Immutable it_{};
Mutable   mt_{};

void Hosuu(size_t x, size_t y)
{
    const array<int, 4> dx{1, 0, -1, 0};
    const array<int, 4> dy{0, 1, 0, -1};
    const Immutable&    it(it_);
    Mutable&            mt(mt_);
    Ref(mt.vec2map, x, y) = 0;
    using XY              = tuple<size_t, size_t>;
    queue<XY> que;
    que.push(make_tuple(x, y));
    while (!que.empty()) {
        const XY& xy(que.front());
        que.pop();
        const int max_count = Get(mt.vec2map, get<0>(xy), get<1>(xy)) + 1;
        for (size_t i = 0; i < dx.size(); ++i) {
            const int nx = get<0>(xy) + dx[i];
            const int ny = get<1>(xy) + dy[i];
            if (0 > nx || nx >= static_cast<int>(it.W) || 0 > ny || ny >= static_cast<int>(it.H)) continue;
            if (Get(it.vec2map, nx, ny) == '#' || Get(mt.vec2map, nx, ny) != Inf) continue;
            que.push(make_tuple(nx, ny));
            Ref(mt.vec2map, nx, ny) = max_count;
            if (mt.max_count < max_count) {
                mt.max_count = max_count;
            }
        }
    }
}

bool MoveG(int count, size_t& x, size_t& y)
{
    Mutable& mt(mt_);
    if (Get(mt.vec2map, x, y) <= count) return true;
    const Immutable& it(it_);
    const size_t     pos = count % it.pattern.size();
    const char       act = it.pattern.at(pos);
    switch (act) {
    case '5':
        // 何もしない
        return false;
    case '8':
        if (y > 0) {
            --y;
        }
        return false;
    case '6':
        if (x < it.W - 1) {
            ++x;
        }
        return false;
    case '4':
        if (x > 0) {
            --x;
        }
        return false;
    case '2':
        if (y < it.H - 1) {
            ++y;
        }
        return false;
    default:
        assert(!"act");
        return false;
    }
}

bool sub()
{
    it_.H = (cin2var<size_t>());
    it_.W = (cin2var<size_t>());
    if (!it_.H && !it_.W) {
        return false;
    }
    it_.vec2map.clear();
    for (size_t i = 0; i < it_.H; ++i) {
        it_.vec2map.push_back(cin2vec<char>(it_.W));
    }
    it_.pattern = cin2var<string>();

    mt_.vec2map = InitVec2<int>(it_.W, it_.H, Inf);

    size_t a_x = 0;
    size_t a_y = 0;
    if (!Find(it_.vec2map, 'A', a_x, a_y)) {
        assert(!"Not found A");
        return false;
    }

    // 移動距離を求める
    Hosuu(a_x, a_y);

    size_t b_x = 0;
    size_t b_y = 0;
    if (!Find(it_.vec2map, 'B', b_x, b_y)) {
        assert(!"Not found B");
        return false;
    }

    // 調査するターン数を決める
    const int  pattern_size = it_.pattern.size();
    const int  g_count      = pattern_size * pattern_size * 2;
    const auto count        = max(mt_.max_count, g_count);

    // ゴーストを動かす
    for (int i = 0; i <= count; ++i) {
        if (MoveG(i, b_x, b_y)) {
            cout << i << ' ' << b_y << ' ' << b_x << endl;
            return true;
        }
    }

    cout << "impossible" << endl;
    return true;
}

int main()
{
    while (true) {
        if (!sub()) {
            break;
        }
    }

    return 0;
}

File: UOA/PCKWU/1046/main.rs
// UOA - PCKWU  1046
// Problem 06: Ghost Buster!

use std::cmp;
use std::collections::VecDeque;

#[allow(dead_code)]
fn read_line(n: usize) -> String {
    let mut ss = String::new();
    for _i in 0..n {
        let mut s = String::new();
        std::io::stdin().read_line(&mut s).unwrap();
        ss.push_str(&s);
    }
    ss
}

macro_rules! parse {
    ($s:expr => $($r:tt)*) => {
        let s = $s;
        let mut iter = s.split_whitespace();
        input_inner!{iter, $($r)*}
    };
    ($($r:tt)*) => {
        use std::io::Read;
        let mut s = String::new();
        std::io::stdin().read_to_string(&mut s).unwrap();
        parse!{s => $($r)*}
    };
}

macro_rules! input_inner {
    ($iter:expr) => {};
    ($iter:expr, ) => {};
    ($iter:expr, $var:ident : $t:tt $($r:tt)*) => {
        let $var = read_value!($iter, $t);
        input_inner!{$iter $($r)*}
    };
    ($iter:expr, mut $var:ident : $t:tt $($r:tt)*) => {
        let mut $var = read_value!($iter, $t);
        input_inner!{$iter $($r)*}
    };
}

macro_rules! read_value {
    ($iter:expr, ( $($t:tt),* )) => {
        ( $(read_value!($iter, $t)),* )
    };
    ($iter:expr, [ $t:tt ; $len:expr ]) => {
        (0..$len).map(|_| read_value!($iter, $t)).collect::<Vec<_>>()
    };
    ($iter:expr, chars) => {
        read_value!($iter, String).chars().collect::<Vec<char>>()
    };
    ($iter:expr, usize1) => {
        read_value!($iter, usize) - 1
    };
    ($iter:expr, $t:ty) => {
        $iter.next().unwrap().parse::<$t>().expect("Parse error")
    };
}

const INF: i32 = i32::max_value();

fn find(tiless: &Vec<Vec<char>>, target: &char) -> (usize, usize) {
    for y in 0..tiless.len() {
        let tiles = &tiless[y];
        for x in 0..tiles.len() {
            let tile = &tiles[x];
            if tile == target {
                return (x, y);
            }
        }
    }
    (0, 0)
}

fn cost(tiless: &Vec<Vec<char>>, xy: &(usize, usize)) -> (Vec<Vec<i32>>, usize) {
    let dx = [1, 0, -1, 0];
    let dy = [0, 1, 0, -1];

    let h = tiless.len() as i32;
    let w = tiless[0].len() as i32;

    let mut costss = Vec::new(); // Vec<Vec<i32>>
    for tiles in tiless {
        let mut costs = Vec::new(); // Vec<i32>
        for _tile in tiles {
            costs.push(INF);
        }
        costss.push(costs);
    }

    costss[xy.1][xy.0] = 0;

    let mut que = VecDeque::new();
    que.push_back(*xy);
    let mut max_count: usize = 0;
    while !que.is_empty() {
        let (x, y) = que.pop_front().expect("error");
        let mc = costss[y][x] + 1;
        for i in 0..dx.len() {
            let nx = (x as i32 + dx[i]) as i32;
            let ny = (y as i32 + dy[i]) as i32;
            if 0 > nx || nx >= w || 0 > ny || ny >= h {
                continue;
            }
            let nx = nx as usize;
            let ny = ny as usize;
            if tiless[ny][nx] == '#' || costss[ny][nx] != INF {
                continue;
            }
            que.push_back((nx, ny));
            costss[ny][nx] = mc;
            if max_count < mc as usize {
                max_count = mc as usize;
            }
        }
    }

    (costss, max_count)
}

fn move_g(
    costss: &Vec<Vec<i32>>,
    i: usize,
    g_moves: &Vec<char>,
    x: &mut usize,
    y: &mut usize,
) -> bool {
    if costss[*y][*x] <= i as i32 {
        return true;
    }
    let h = costss.len();
    let w = costss[0].len();
    let pos = i % g_moves.len();
    let act = g_moves[pos];
    match act {
        '5' => {}
        '8' => {
            if *y > 0 {
                *y -= 1;
            }
        }
        '6' => {
            if *x < w - 1 {
                *x += 1;
            }
        }
        '4' => {
            if *x > 0 {
                *x -= 1;
            }
        }
        '2' => {
            if *y < h - 1 {
                *y += 1;
            }
        }
        _ => {}
    }
    false
}

fn sub() -> bool {
    parse! {
        read_line(1) =>
        h: usize,
        w: usize,
    }
    if h == 0 && w == 0 {
        return false;
    }
    parse! {
        read_line(h) =>
        tiless: [chars; h],
    }
    parse! {
        read_line(1) =>
        g_moves: chars,
    }

    // 少女と幽霊を探す
    let a = find(&tiless, &'A');
    let b = find(&tiless, &'B');

    // 移動距離を求める
    let (costss, max_count) = cost(&tiless, &a);

    // 調査するターン数を決める
    let g_move_len = g_moves.len();
    let g_count = g_move_len * g_move_len * 2;
    let count = cmp::max(max_count, g_count) + 1;

    // 幽霊を動かす
    let mut x = b.0;
    let mut y = b.1;
    for i in 0..count {
        if move_g(&costss, i, &g_moves, &mut x, &mut y) {
            println!("{} {} {}", i, y, x);
            return true;
        }
    }

    println!("impossible");
    true
}

fn main() {
    while sub() {}
}
